File: leetcode/1001-2000/1701-1800/1751-1760/1752-check-if-array-is-sorted-and-rotated/CheckIfArrayIsSortedAndRotated.ts
function check(nums: number[]): boolean {
  const n = nums.length;

  if (n < 3) {
    return true;
  }

  let i = 1;

  while (i < n) {
    if (nums[i] < nums[i - 1]) {
      break;
    }

    i++;
  }

  if (i === n) {
    return true;
  }

  i++;

  while (i < n) {
    if (nums[i] < nums[i - 1]) {
      return false;
    }

    i++;
  }

  return nums[n - 1] <= nums[0];
};

console.log(check([3, 4, 5, 1, 2]));
console.log(check([2, 1, 3, 4]));

File: leetcode/1001-2000/1701-1800/1751-1760/1752-check-if-array-is-sorted-and-rotated/CheckIfArrayIsSortedAndRotated.java
public class CheckIfArrayIsSortedAndRotated {
  public static void main(String[] args) {
    System.out.println(new Solution().check(new int[] { 3, 1, 2, 3, 4 }));
  }
}

class Solution {
  public boolean check(int[] nums) {
    if (nums.length < 3) {
      return true;
    }

    int obstacles = 0;

    if (nums[0] < nums[nums.length - 1]) {
      obstacles++;
    }

    for (int i = 1; i < nums.length; i++) {
      if (nums[i] < nums[i - 1]) {
        obstacles++;

        if (obstacles > 1) {
          return false;
        }
      }
    }

    return true;
  }
}
